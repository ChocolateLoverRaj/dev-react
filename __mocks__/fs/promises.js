/* global jest */

// Mocker for fs/promises

const fs = jest.createMockFromModule('fs/promises')

const EventEmitter = require('eventemitter3')

const frozen = new Set()
fs._frozen = frozen
const errorFiles = new Set()
fs._errorFiles = errorFiles
const files = new Map()
fs._files = files
fs._reset = () => {
  frozen.clear()
  errorFiles.clear()
  files.clear()
}

const mock = new EventEmitter()
  .on('freeze', filename => {
    frozen.add(filename)
  })
  .on('unfreeze', filename => {
    frozen.delete(filename)
  })
  .on('throw', filename => {
    errorFiles.add(filename)
  })
  .on('reset', () => {
    fs._reset()
  })
fs._mock = mock

const onceUnfrozen = filename => new Promise((resolve, reject) => {
  const handler = f => {
    if (f === filename) {
      mock.off(handler)
      resolve()
    }
  }
  mock.on('unfreeze', handler)
})

fs.writeFile = async (filename, content) => {
  if (errorFiles.has(filename)) {
    throw new Error('Error writing file.')
  }
  const write = () => {
    files.set(filename, content)
    mock.emit('wrote', filename, content)
  }
  if (frozen.has(filename)) {
    await onceUnfrozen(filename)
    write()
  } else {
    write()
  }
}

fs.readFile = async filename => {
  if (errorFiles.has(filename)) {
    throw new Error('Error reading file.')
  }
  const read = () => {
    mock.emit('read', filename)
    return files.get(filename)
  }
  if (frozen.has(filename)) {
    await onceUnfrozen(filename)
    return read()
  } else {
    return read()
  }
}

fs.unlink = async filename => {
  files.delete(filename)
  mock.emit('unlinked', filename)
}

module.exports = fs
