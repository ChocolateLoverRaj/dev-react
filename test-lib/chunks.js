const chunks = stream => new Promise((resolve, reject) => {
  const chunks = []
  stream.on('data', data => {
    chunks.push(data)
  })
  stream.on('end', () => {
    resolve(chunks)
  })
  stream.on('error', reject)
})

module.exports = chunks
