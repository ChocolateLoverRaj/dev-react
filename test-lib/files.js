export class File {
  constructor () {
    this.canRead = true
    this.canWrite = true
  }
}

export class NormalFile extends File {
  constructor (content) {
    super()
    this.content = content
  }
}

export class Dir extends File {
  constructor () {
    super()
    this.files = new Map()
  }
}

class EnoentError extends Error {
  constructor () {
    super(...arguments)
    this.code = 'ENOENT'
  }
}

export const topDir = new Dir()

export const reset = () => {
  topDir.files.clear()
}

export const getFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
  return file
}

export const setFile = (path, fileToSet) => {
  let dir = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (fileTree.length === filenames.length) {
      dir.files.set(filename, fileToSet)
      return
    }
    if (dir.files.has(filename)) {
      dir = dir.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}

export const unlinkFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      if (fileTree.length === filenames.length) {
        file.files.delete(filename)
        return
      }
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}
