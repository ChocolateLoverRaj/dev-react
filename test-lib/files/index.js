// TODO: structure with folders and files
const topDir = new Map()

export default topDir