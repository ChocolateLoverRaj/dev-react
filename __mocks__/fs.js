// Mocker for fs
import { getFile } from '../test-lib/files.js'
import tick from '../test-lib/tick.js'
import { Readable, PassThrough } from 'stream'

export const _readStreamErrorFiles = new Set()

export const _reset = () => {
  _readStreamErrorFiles.clear()
}

export const createReadStream = path => {
  if (_readStreamErrorFiles.has(path)) {
    const stream = new PassThrough()
    tick().then(() => {
      stream.destroy(new Error('Error creating a read stream.'))
    })
    return stream
  }
  return Readable.from(getFile(path).content.split(''))
}
