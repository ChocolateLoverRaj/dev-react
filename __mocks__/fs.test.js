/* global afterEach, test, expect */

import { createReadStream, _readStreamErrorFiles, _reset } from './fs.js'
import files from '../test-lib/files.js'
import chunks from '../test-lib/chunks.js'

afterEach(() => {
  files.clear()
})
afterEach(_reset)

test('createReadStream', async () => {
  files.set('file', 'hi')
  expect(await chunks(createReadStream('file'))).toStrictEqual(['h', 'i'])
})

test('_readStreamErrorFiles', async () => {
  _readStreamErrorFiles.add('file')
  await expect(chunks(createReadStream('file'))).rejects.toMatchSnapshot()
})

test('_reset', () => {
  _readStreamErrorFiles.add('file')
  _reset()
  expect(_readStreamErrorFiles.size).toBe(0)
})

test('non-existent file', async () => {
  expect(() => {
    createReadStream('file')
  }).toThrowErrorMatchingSnapshot()
})
