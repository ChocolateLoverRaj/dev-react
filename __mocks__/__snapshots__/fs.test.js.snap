// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_readStreamErrorFiles 1`] = `[Error: Error creating a read stream.]`;

exports[`non-existent file 1`] = `"File: file does not exist"`;
